File: nanopay/src/net/nanopay/cico/spi/alterna/AlternaSFTPService.java
package net.nanopay.cico.spi.alterna;

import com.jcraft.jsch.ChannelSftp;
import com.jcraft.jsch.JSch;
import com.jcraft.jsch.Session;
import foam.core.ContextAwareSupport;
import foam.lib.json.OutputterMode;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.util.Date;
import java.util.Properties;

public class AlternaSFTPService
    extends ContextAwareSupport
    implements SFTPService
{
  protected final String HOST = "ftp.eftcanada.com";
  protected final int PORT = 22;
  protected final String USER = "eftcadtest1";
  protected final String PASSWORD = "nLGlp8Du";
  protected final String WORKING_DIR = "/";

  @Override
  public void sendCICOFile() {
    Date now = new Date();
    ByteArrayOutputStream baos = new ByteArrayOutputStream();
    CsvUtil.writeCsvFile(getX(), baos, OutputterMode.STORAGE);

    Session session = null;
    ChannelSftp channel = null;

    try {
      // create session with user name and password
      JSch jsch = new JSch();
      session = jsch.getSession(USER, HOST, PORT);
      session.setPassword(PASSWORD);

      // add configuration
      Properties config = new java.util.Properties();
      config.put("StrictHostKeyChecking", "no");
      session.setConfig(config);
      session.connect();

      // open SFTP connection and upload file
      channel = (ChannelSftp) session.openChannel("sftp");
      channel.connect();
      channel.cd(WORKING_DIR);
      channel.put(new ByteArrayInputStream(baos.toByteArray()), CsvUtil.generateFilename(now));
      channel.exit();
    } catch ( Exception e ) {
      e.printStackTrace();
    } finally {
      // close channels
      if ( channel != null ) channel.disconnect();
      if ( session != null ) session.disconnect();
      System.out.println("Host Session disconnected.");
    }
  }

  @Override
  public void start() {

  }
}

File: nanopay/src/net/nanopay/cico/spi/alterna/AlternaSFTPService.js
foam.CLASS({
  package: 'net.nanopay.cico.spi.alterna',
  name: 'AlternaSFTPService',

  implements: [
    'net.nanopay.cico.spi.alterna.SFTPService'
  ],

  javaImports: [
    'com.jcraft.jsch.Channel',
    'com.jcraft.jsch.ChannelSftp',
    'com.jcraft.jsch.JSch',
    'com.jcraft.jsch.Session',
    'foam.lib.json.OutputterMode',
    'foam.core.X',
    'foam.nanos.logger.Logger',
    'java.io.ByteArrayInputStream',
    'java.io.ByteArrayOutputStream',
    'java.util.Date',
    'java.util.Properties',
    'java.util.Vector',
    'net.nanopay.cico.model.EFTReturnFileCredentials'
],

  methods: [
    {
      name: 'sendCICOFile',
      javaCode:
        `Date now = new Date();
ByteArrayOutputStream baos = new ByteArrayOutputStream();

X x = getX();
EFTReturnFileCredentials credentials = (EFTReturnFileCredentials) x.get("ETFReturnFileCredentials");
CsvUtil.writeCsvFile(x, baos, OutputterMode.STORAGE);

final Logger logger = (Logger) x.get("logger");
Session session = null;
Channel channel = null;
ChannelSftp channelSftp;
try {
  // create session with user name and password
  JSch jsch = new JSch();
  session = jsch.getSession(credentials.getUser(), credentials.getHost(), credentials.getPort());
  session.setPassword(credentials.getPassword());
  
  // add configuration
  Properties config = new Properties();
  config.put("StrictHostKeyChecking", "no");
  session.setConfig(config);
  session.connect();
  
  // open SFTP connection and upload file
  channel = session.openChannel("sftp");
  channel.connect();
  channelSftp = (ChannelSftp) channel;

  channelSftp.cd("/");
  
  String filename = CsvUtil.generateFilename(now);
  
  Vector list = channelSftp.ls("/Archive/");
  boolean csvFileExist = false;
  
  for ( Object entry : list ) {
    ChannelSftp.LsEntry e = (ChannelSftp.LsEntry) entry;
    if ( e.getFilename().equals(filename) ) {
      csvFileExist = true;
    }
  }
  
  if ( ! csvFileExist ) {
    channelSftp.put(new ByteArrayInputStream(baos.toByteArray()), filename);
    logger.debug("CICO CSV file sent");
  } else {
    logger.warning("duplicate csv file not sent", filename, System.getProperty("user.name"));
  } 
  
  channelSftp.exit();
} catch ( Exception e ) {
  logger.error(e);
} finally {
  // close channels
  if ( channel != null ) channel.disconnect();
  if ( session != null ) session.disconnect();
}`
    }
  ]
});
